File: python/search_a_2d_matrix.py
class Solution:
    def searchMatrix(self, matrix: list[list[int]], target: int) -> bool:
        mertedList: list[int] = []
        for m in matrix:
            mertedList.extend(m)

        l: int = 0
        r: int = len(mertedList) - 1
        while l <= r:
            mid: int = int((l + r) / 2)
            if mertedList[mid] == target:
                return True
            elif mertedList[mid] > target:
                r = mid - 1
            elif mertedList[mid] < target:
                l = mid + 1

        return False

    def searchMatrix2(self, matrix: list[list[int]], target: int) -> bool:
        """tiem complexity: O(log(m*n))"""
        top: int = 0
        bot: int = len(matrix) - 1
        target_row: int = 0
        while top <= bot:
            mid_row: int = (top + bot) // 2
            if matrix[mid_row][0] <= target and target <= matrix[mid_row][-1]:
                target_row = mid_row
                break
            if matrix[mid_row][0] > target:
                bot = mid_row - 1
            elif matrix[mid_row][-1] < target:
                top = mid_row + 1
        if top > bot:
            return False
        left: int = 0
        right: int = len(matrix[target_row]) - 1
        while left <= right:
            mid: int = (left + right) // 2
            if matrix[target_row][mid] == target:
                return True
            if matrix[target_row][mid] > target:
                right = mid - 1
            elif matrix[target_row][mid] < target:
                left = mid + 1
        return False


def test_searchMatrix_case_1():
    # arrange
    matrix: list[list[int]] = [[1, 3, 5, 7], [10, 11, 16, 20], [23, 30, 34, 60]]
    target: int = 3
    expected: bool = True

    # act
    solution = Solution()
    actual = solution.searchMatrix(matrix, target)

    # assert
    assert actual == expected


def test_searchMatrix_case_2():
    # arrange
    matrix = [[1, 3, 5, 7], [10, 11, 16, 20], [23, 30, 34, 60]]
    target = 13
    expected = False

    # act
    solution = Solution()
    actual = solution.searchMatrix(matrix, target)

    # assert
    assert actual == expected


def test_searchMatrix_case_3():
    # arrange
    matrix = [[1, 3]]
    target = 3
    expected = True

    # act
    solution = Solution()
    actual = solution.searchMatrix(matrix, target)

    # assert
    assert actual == expected


def test_searchMatrix_case_4():
    # arrange
    matrix: list[list[int]] = [[1, 3, 5, 7], [10, 11, 16, 20], [23, 30, 34, 60]]
    target: int = 3
    expected: bool = True

    # act
    solution = Solution()
    actual = solution.searchMatrix2(matrix, target)

    # assert
    assert actual == expected


def test_searchMatrix_case_5():
    # arrange
    matrix = [[1, 3, 5, 7], [10, 11, 16, 20], [23, 30, 34, 60]]
    target = 13
    expected = False

    # act
    solution = Solution()
    actual = solution.searchMatrix2(matrix, target)

    # assert
    assert actual == expected


def test_searchMatrix_case_6():
    # arrange
    matrix = [[1, 3]]
    target = 3
    expected = True

    # act
    solution = Solution()
    actual = solution.searchMatrix2(matrix, target)

    # assert
    assert actual == expected

File: rust/src/code/search_a_2d_matrix.rs
struct Solution {}

impl Solution {
    pub fn search_matrix(matrix: Vec<Vec<i32>>, target: i32) -> bool {
        let mut merged_vector: Vec<i32> = Vec::new();
        for v in matrix {
            merged_vector.extend(v);
        }
        let mut l = 0;
        let mut r = merged_vector.len() - 1;
        while l <= r {
            let mid = (l + r) / 2;
            if merged_vector[mid] == target {
                return true;
            } else if merged_vector[mid] > target {
                if mid == 0 {
                    break;
                }
                r = mid - 1;
            } else {
                l = mid + 1;
            }
        }

        false
    }

    pub fn search_matrix2(matrix: Vec<Vec<i32>>, target: i32) -> bool {
        let mut bot = 0;
        let mut top = matrix.len() - 1;
        let mut l = 0;
        let mut r = matrix[0].len() - 1;

        while bot <= top {
            let mid = (bot + top) / 2;
            if target > matrix[mid][r] {
                bot = mid + 1;
            } else if target < matrix[mid][0] {
                if mid > 0 {
                    top = mid - 1;
                } else {
                    break;
                }
            } else {
                break;
            }
        }

        if bot > top {
            return false;
        }
        let row = (bot + top) / 2;

        while l <= r {
            let mid = (l + r) / 2;
            if target == matrix[row][mid] {
                return true;
            } else if target > matrix[row][mid] {
                l = mid + 1;
            } else if target < matrix[row][mid] {
                if mid > 0 {
                    r = mid - 1;
                } else {
                    break;
                }
            }
        }

        return false;
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn search_matrix_case_1() {
        // arrange
        let matrix = vec![vec![1, 3, 5, 7], vec![10, 11, 16, 20], vec![23, 30, 34, 60]];
        let target = 3;
        let expected = true;

        // act
        let actual = Solution::search_matrix(matrix, target);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn search_matrix_case_2() {
        // arrange
        let matrix = vec![vec![1, 3, 5, 7], vec![10, 11, 16, 20], vec![23, 30, 34, 60]];
        let target = 13;
        let expected = false;

        // act
        let actual = Solution::search_matrix(matrix, target);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn search_matrix_case_3() {
        // arrange
        let matrix = vec![vec![0]];
        let target = 0;
        let expected = true;

        // act
        let actual = Solution::search_matrix(matrix, target);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn search_matrix_case_4() {
        // arrange
        let matrix = vec![vec![1]];
        let target = 0;
        let expected = false;

        // act
        let actual = Solution::search_matrix(matrix, target);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn search_matrix_case_5() {
        // arrange
        let matrix = vec![vec![1, 3, 5, 7], vec![10, 11, 16, 20], vec![23, 30, 34, 60]];
        let target = 3;
        let expected = true;

        // act
        let actual = Solution::search_matrix2(matrix, target);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn search_matrix_case_6() {
        // arrange
        let matrix = vec![vec![1, 3, 5, 7], vec![10, 11, 16, 20], vec![23, 30, 34, 60]];
        let target = 13;
        let expected = false;

        // act
        let actual = Solution::search_matrix2(matrix, target);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn search_matrix_case_7() {
        // arrange
        let matrix = vec![vec![0]];
        let target = 0;
        let expected = true;

        // act
        let actual = Solution::search_matrix2(matrix, target);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn search_matrix_case_8() {
        // arrange
        let matrix = vec![vec![1]];
        let target = 0;
        let expected = false;

        // act
        let actual = Solution::search_matrix2(matrix, target);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn search_matrix_case_9() {
        // arrange
        let matrix = vec![vec![1]];
        let target = 2;
        let expected = false;

        // act
        let actual = Solution::search_matrix2(matrix, target);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn search_matrix_case_10() {
        // arrange
        let matrix = vec![vec![1], vec![3], vec![5]];
        let target = 3;
        let expected = true;

        // act
        let actual = Solution::search_matrix2(matrix, target);

        // assert
        assert_eq!(expected, actual);
    }
}
